Objects/memoryobject.c

```diff
@@ -3160,6 +3161,48 @@ static PyMethodDef memory_methods[] = {
     {NULL,          NULL}
 };
 
+/////////////////////////////////////////////////////////////////////////////
+
+static PyObject *memory_iter(PyObject *seq);
+
+typedef struct {
+    PyObject_HEAD
+    Py_ssize_t it_index;
+    PyMemoryViewObject *it_seq; // Set to NULL when iterator is exhausted
+} memoryiterobject;
+
+
+static PyTypeObject PyMemoryIter_Type = {
+    PyVarObject_HEAD_INIT(&PyType_Type, 0)
+    .tp_name = "memory_iterator",
+    .tp_basicsize = sizeof(memoryiterobject),
+    // methods
+    .tp_getattro = PyObject_GenericGetAttr,
+    .tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC,
+    .tp_iter = PyObject_SelfIter,
```

Comment:
Looks like you need a few more methods; at least `tp_dealloc` and `tp_iternext`. You can copy `tp_dealloc` from `bytesobject.c`; `tp_iternext` is the beef of the work (though for 1d bytes view it can be similar to `bytesobject.c`).

You also need `tp_traverse`, which again you can copy from `bytesobject.c`.

Reply:
I forgot to add tp_dealloc and was leaving tp_iternext for the next PR as I was aware that it would take a bit more work to get right.

On a tangent note, I was wondering if it would make sense to avoid all this copying, as all these functions are basically the same between the different modules (bytesobject.c, bytearrayobject.c, memoryobject.c).
Wouldn't it be possible to write all these similar functions in a more general form (i.e having an extra parameter passed in with the _iteratorobject_ type to be used) so we would only need one <iterator>_traverse, <iterator>_dealloc, etc?

Reply:
> On a tangent note, I was wondering if it would make sense to avoid all this copying, as all these functions are basically the same between the different modules (bytesobject.c, bytearrayobject.c, memoryobject.c).
> Wouldn't it be possible to write all these similar functions in a more general form (i.e having an extra parameter passed in with the _iteratorobject_ type to be used) so we would only need one _traverse, _dealloc, etc?

I suppose a different architecture could have been selected early on, but it's too late for such a big change now. We just have to grin and bear it. (It is a little simpler if you don't want to support GC, but it's actually possible to imagine this to be part of a cycle of object references, so we should support it, like you started doing.

A few things could be done with macros or helper functions, but fundamentally the `tp_iter` protocol doesn't know the layout of the specific iterator object -- in particular it doesn't know that there is an `it_seq` field receiving a reference to the original object.

So you just have to grin and bear it! That's working on CPython for you.

Reply:
Got it. Thank you for the clear explanation. I will continue to work on the code.